[MIR] Support save/restore points with independent sets of registers

enoskova-sc opened the pull request:
This patch adds the MIR parsing and serialization support for save and restore points with subsets of callee saved registers.  That is, it syntactically allows a function to contain two or more distinct sub-regions in which distinct subsets of registers are spilled/filled as callee save.  This is useful if e.g. one of the CSRs isn't modified in one of the sub-regions, but is in the other(s).  

Support for actually using this capability in code generation is still forthcoming.  This patch is the next logical step for multiple save/restore points support. 

All points are now stored in DenseMap from MBB to vector of CalleeSavedInfo. 

Shrink-Wrap points split Part 4.
RFC: https://discourse.llvm.org/t/shrink-wrap-save-restore-points-splitting/83581

Part 1: https://github.com/llvm/llvm-project/pull/117862 (landed)
Part 2: https://github.com/llvm/llvm-project/pull/119355 (landed)
Part 3: https://github.com/llvm/llvm-project/pull/119357 (landed)
Part 5: https://github.com/llvm/llvm-project/pull/119359 (likely to be further split)

enoskova-sc commented:
@topperc, @arsenm, could you take a look, please.

topperc commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -809,21 +847,105 @@ class MachineFrameInfo {
+
+  void setSavePoint(MachineBasicBlock *MBB, std::vector<Register> &Regs) {
+    if (SavePoints.contains(MBB))
+      SavePoints[MBB] = Regs;
```
Why do we need to check `contains`? Can't we just always use `SavePoints[MBB] = Regs`?

topperc commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -809,21 +847,105 @@ class MachineFrameInfo {
+    RestorePoints = std::move(NewRestorePoints);
+  }
+
+  void setSavePoint(MachineBasicBlock *MBB, std::vector<Register> &Regs) {
```
`const std::vector<Register> &Regs`

topperc commented on llvm/include/llvm/CodeGen/MachineFrameInfo.h:
```diff
@@ -809,21 +847,105 @@ class MachineFrameInfo {
+  }
+
+  void setRestorePoint(MachineBasicBlock *MBB, std::vector<Register> &Regs) {
+    if (RestorePoints.contains(MBB))
```
Why do we need to check contains?

topperc commented on llvm/lib/CodeGen/MIRParser/MIRParser.cpp:
```diff
@@ -1058,8 +1053,40 @@ bool MIRParserImpl::initializeConstantPool(PerFunctionMIParsingState &PFS,
+    if (parseMBBReference(PFS, MBB, MBBSource.Value))
+      return true;
+
+    std::vector<Register> Registers{};
```
No need for the curly braces

enoskova-sc replied:
addressed

michaelmaitland commented:
Does this PR have the right commits? I think it might have the same commits as part 5.

enoskova-sc committed:
[MIR] Support save/restore points with independent sets of registers

This patch adds the MIR parsing and serialization support
for save and restore points with subsets of callee saved registers.
It syntactically allows a function to contain two or more distinct sub-regions
with distinct subsets of registers spilled/filled as callee save.
It can be useful, when one of the CSRs isn't modified in one of the sub-regions,
but is in the other(s).

Shrink-Wrap points split. Part 4.
RFC: https://discourse.llvm.org/t/shrink-wrap-save-restore-points-splitting/83581

enoskova-sc commented:
@preames, I split the 4th patch, and this part introduces information about registers spilled/restored in save/restore points (as you suggested), but it still does not provide the possibility to have multiple save/restore points.

enoskova-sc commented:
@preames, @ppenzin, @arsenm, @topperc, do you have any comments?

enoskova-sc commented:
@preames, do you still have any doubts, concerning shrink-wrap improvement, proposed by me?
If do, may be it is worth discussing. 
If not, please, take a look at my MR, when you have time.

preames commented:
> @preames, do you still have any doubts, concerning shrink-wrap improvement, proposed by me?
> If do, may be it is worth discussing.

I wouldn't call them "doubts", more questions about the path.  After this, is your intention to post a patch which goes from MIR to correct codegen?  And then after that a patch which heuristically inserts the distinct save/restore?  (I see that your final patch contains all of this, but it's hard to parse apart.)

enoskova-sc commented:
>  After this, is your intention to post a patch which goes from MIR to correct codegen? And then after that a patch which heuristically inserts the distinct save/restore? (I see that your final patch contains all of this, but it's hard to parse apart.)

After this patch I plan to support multiple points in RISC-V backend (may be in two patches, if I manage to split). And then, yes, I'll move to shrink-wrap pass and will modify its algorithm so that it can produce multiple points.

preames commented on llvm/lib/CodeGen/ShrinkWrap.cpp:
```diff
@@ -967,12 +967,12 @@ bool ShrinkWrapImpl::run(MachineFunction &MF) {
+
+  // List of CalleeSavedInfo for registers will be added during prologepilog
+  // pass
+  SaveRestorePoints SavePoints({{Save, {}}});
```
Just to confirm, is the invariant here that Save was always non-null along this path?  If so, the change looks okay, otherwise you're adding a null block.  

(To be clear, my LGTM is conditional on the answer to this question being always non-null.)

preames reviewed (approved):
LGTM conditional on answer to inline question.
